File: src/enigma.swift
// Enigma Machine — Rosetta Code Reference (Swift)
// Wehrmacht Enigma I: 3 rotors, Reflector B, plugboard.
// PeopleTec Inc. — Guinness World Record Attempt 2026

let FWD = ["EKMFLGDQVZNTOWYHXUSPAIBRCJ","AJDKSIRUXBLHWTMCQGZNPYFVOE","BDFHJLCPRTXVZNYEIWGAKMUSQO"]
let BWD = ["UWYGADFPVZBECKMTHXSLRINQOJ","AJPCZWRLFBDKOTYUQGENHXMIVS","TAGBPCSDQEUFVNZHYIXJWLRKOM"]
let NOTCH: [Int] = [Int(Character("Q").asciiValue!)-65, Int(Character("E").asciiValue!)-65, Int(Character("V").asciiValue!)-65]
let REFLECTOR = Array("YRUHQSLDPXNGOKMIEBFZCWVJAT")

func idx(_ c: Character) -> Int { Int(c.asciiValue!) - 65 }
func chr(_ i: Int) -> Character { Character(UnicodeScalar(i + 65)!) }

struct Enigma {
    var id: [Int], off: [Int], plug: [Int]

    init(_ rotors: [Int], _ key: String, _ plugboard: [String] = []) {
        id = rotors
        off = Array(key).map { idx($0) }
        plug = Array(0..<26)
        for p in plugboard {
            let chars = Array(p)
            let (a, b) = (idx(chars[0]), idx(chars[1]))
            plug[a] = b; plug[b] = a
        }
    }

    func fwd(_ r: Int, _ i: Int) -> Int {
        let c = (i + off[r]) % 26
        return (idx(Array(FWD[id[r]])[c]) - off[r] + 26) % 26
    }
    func bwd(_ r: Int, _ i: Int) -> Int {
        let c = (i + off[r]) % 26
        return (idx(Array(BWD[id[r]])[c]) - off[r] + 26) % 26
    }

    mutating func step() {
        if off[1] == NOTCH[id[1]] { off[1]=(off[1]+1)%26; off[0]=(off[0]+1)%26 }
        else if off[2] == NOTCH[id[2]] { off[1]=(off[1]+1)%26 }
        off[2]=(off[2]+1)%26
    }

    mutating func pressKey(_ ch: Character) -> Character {
        step()
        var i = idx(ch)
        i = plug[i]
        i = fwd(2,i); i = fwd(1,i); i = fwd(0,i)
        i = idx(REFLECTOR[i])
        i = bwd(0,i); i = bwd(1,i); i = bwd(2,i)
        i = plug[i]
        return chr(i)
    }

    mutating func encrypt(_ text: String) -> String {
        String(text.uppercased().filter{$0.isLetter}.map{pressKey($0)})
    }
}

let tests: [([Int],String,[String],String,String)] = [
    ([0,1,2],"AAA",[],"AAAAA","BDZGO"),
    ([0,1,2],"AAA",[],"HELLOWORLD","ILBDAAMTAZ"),
    ([0,1,2],"AAA",[],"ATTACKATDAWN","BZHGNOCRRTCM"),
    ([0,1,2],"MCK",[],"HELLOWORLD","DLTBBQVPQV"),
    ([2,0,1],"AAA",[],"HELLOWORLD","KZHDFQYHXT"),
    ([0,1,2],"AAA",["AB","CD","EF"],"HELLOWORLD","IKACBBMTBF"),
]

var allOk = true
for (t,(r,k,p,pt,exp)) in tests.enumerated() {
    var e = Enigma(r,k,p)
    let ct = e.encrypt(pt)
    let ok = ct == exp
    print("Test \(t+1): \(pt.padding(toLength:20,withPad:" ",startingAt:0)) -> \(ct.padding(toLength:15,withPad:" ",startingAt:0)) [\(ok ? "PASS" : "FAIL")]")
    if !ok { allOk = false }
}
print(allOk ? "\nALL 6 TESTS PASSED" : "\nSOME TESTS FAILED")

File: src/enigma.kt
// Enigma Machine — Rosetta Code Reference (Kotlin)
// Wehrmacht Enigma I: 3 rotors, Reflector B, plugboard.
// PeopleTec Inc. — Guinness World Record Attempt 2026

val FWD = arrayOf("EKMFLGDQVZNTOWYHXUSPAIBRCJ","AJDKSIRUXBLHWTMCQGZNPYFVOE","BDFHJLCPRTXVZNYEIWGAKMUSQO")
val BWD = arrayOf("UWYGADFPVZBECKMTHXSLRINQOJ","AJPCZWRLFBDKOTYUQGENHXMIVS","TAGBPCSDQEUFVNZHYIXJWLRKOM")
val NOTCH = charArrayOf('Q','E','V')
val REFLECTOR = "YRUHQSLDPXNGOKMIEBFZCWVJAT"

class Enigma(rotors: IntArray, key: String, plugboard: List<String> = emptyList()) {
    private val id = rotors
    private val off = intArrayOf(key[0]-'A', key[1]-'A', key[2]-'A')
    private val plug = IntArray(26) { it }.also { p ->
        plugboard.forEach { s -> val a=s[0]-'A'; val b=s[1]-'A'; p[a]=b; p[b]=a }
    }

    private fun fwd(r:Int,i:Int):Int { val c=(i+off[r])%26; return(FWD[id[r]][c]-'A'-off[r]+26)%26 }
    private fun bwd(r:Int,i:Int):Int { val c=(i+off[r])%26; return(BWD[id[r]][c]-'A'-off[r]+26)%26 }

    private fun step() {
        if (off[1]+'A'.code == NOTCH[id[1]].code) { off[1]=(off[1]+1)%26; off[0]=(off[0]+1)%26 }
        else if (off[2]+'A'.code == NOTCH[id[2]].code) { off[1]=(off[1]+1)%26 }
        off[2]=(off[2]+1)%26
    }

    fun pressKey(ch: Char): Char {
        step(); var i=ch-'A'
        i=plug[i]
        i=fwd(2,i); i=fwd(1,i); i=fwd(0,i)
        i=REFLECTOR[i]-'A'
        i=bwd(0,i); i=bwd(1,i); i=bwd(2,i)
        i=plug[i]
        return 'A'+i
    }

    fun encrypt(text:String):String = text.uppercase().filter{it in 'A'..'Z'}.map{pressKey(it)}.joinToString("")
}

fun main() {
    data class TV(val r:IntArray, val k:String, val p:List<String>, val pt:String, val exp:String)
    val tests = listOf(
        TV(intArrayOf(0,1,2),"AAA",emptyList(),"AAAAA","BDZGO"),
        TV(intArrayOf(0,1,2),"AAA",emptyList(),"HELLOWORLD","ILBDAAMTAZ"),
        TV(intArrayOf(0,1,2),"AAA",emptyList(),"ATTACKATDAWN","BZHGNOCRRTCM"),
        TV(intArrayOf(0,1,2),"MCK",emptyList(),"HELLOWORLD","DLTBBQVPQV"),
        TV(intArrayOf(2,0,1),"AAA",emptyList(),"HELLOWORLD","KZHDFQYHXT"),
        TV(intArrayOf(0,1,2),"AAA",listOf("AB","CD","EF"),"HELLOWORLD","IKACBBMTBF"),
    )
    var allOk = true
    tests.forEachIndexed { t, tv ->
        val ct = Enigma(tv.r, tv.k, tv.p).encrypt(tv.pt)
        val ok = ct == tv.exp
        println("Test ${t+1}: ${tv.pt.padEnd(20)} -> ${ct.padEnd(15)} [${if(ok)"PASS" else "FAIL"}]")
        if (!ok) allOk = false
    }
    println(if (allOk) "\nALL 6 TESTS PASSED" else "\nSOME TESTS FAILED")
}
